server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -1048,4 +1099,35 @@ static List<Diagnosis.Resource> getRestoreFromSnapshotAffectedResources(
             return affectedResources;
         }
     }
+
+    static class SearchableSnapshotsState {
+        private final Set<String> searchableSnapshotWithUnavailableShard = new HashSet<>();
+        private final Set<String> searchableSnapshotWithOriginalIndexAvailable = new HashSet<>();
+
+        void addSearchableSnapshotWithUnavailableShard(String indexName) {
+            searchableSnapshotWithUnavailableShard.add(indexName);
+        }
+
+        void addSearchableSnapshotWithOriginalIndexAvailable(String indexName) {
+            searchableSnapshotWithOriginalIndexAvailable.add(indexName);
+        }
+
+        Set<String> getRedSearchableSnapshots() {
+            return Sets.difference(searchableSnapshotWithUnavailableShard, searchableSnapshotWithOriginalIndexAvailable);
+        }
+
+        // If the original index of a searchable snapshot with unavailable shards is available then we remove the searchable snapshot
+        // from the list of the unavailable searchable snapshots because the data is available via the original index.
+        void updateSearchableSnapshotWithAvailableIndices(Metadata clusterMetadata, Set<String> indicesWithUnavailableShards) {
+            for (String index : searchableSnapshotWithUnavailableShard) {
+                Settings indexSettings = clusterMetadata.index(index).getSettings();
+                String originalIndex = indexSettings.get(SearchableSnapshotsSettings.SEARCHABLE_SNAPSHOT_INDEX_NAME_SETTING_KEY);
```

Comment:
I can add it, no objections, but out of curiosity, what is it offering? Theoretically, we would not be able to get here if it was null, because we need the settings to determine if it's a searchable snapshot.

Reply:
I'm a bit concerned about this line `clusterMetadata.index(index)` that could yield a `null` `IndexMetadata`.
It's that metadata that I'd like to assert it's not null (sorry my phrasing around the settings before was not precise enough). 

The assertion will give us a bit of protection against misusing the `updateSearchableSnapshotWithAvailableIndices` method with a different `Metadata` to the one we inspected the indices settings before.

Reply:
I will add it. No objections.